## Content

  - **Installation of pandas**
    - Importing pandas
    - Importing the dataset
    - Dataframe/Series
  
  - **Basic ops on a DataFrame**
    - df.info()
    - df.head()
    - df.tail()
    - df.shape()
  - **Creating Dataframe from Scratch**

  - **Basic ops on columns**
    - Different ways of accessing cols
    - Check for Unique values
    - Rename column
    - Deleting col
    - Creating new cols




### Today's Agenda
- Today's lecture is about **Pandas** library
- We'll see **what** is Pandas
- **Why** we use this library
- We'll also look at some **interesting tasks** we can do **using Pandas**


<!-- ### Before we start: [00:10 - 00:15]

- Download .csv files from shared link


- Create a folder in same directory as your running Jupyter Notebook


- Rename this folder to **data**


- **Place the downloaded .csv files in that folder called "data"**


- This way, we **don't have to provide complete path** every time we want to access any of our .csv files


- We just have to provide the **relative path**
    - `data/<csv_file_name>` -->



## Installing Pandas

In [ ]:
# !pip install pandas

## Importing Pandas

- You should be able to import Pandas after installing it


- We'll import `pandas` as its **alias name `pd`**

In [1]:
import pandas as pd
import numpy as np

## Introduction: Why to use Pandas?


#### How is it different from numpy ?

  - The major **limitation of numpy** is that it can only work with 1 datatype at a time

  - Most real-world datasets contain a mixture of different datatypes
    - Like **names of places would be string** but their **population would be int**
  
==> It is **difficult to work** with data having **heterogeneous values using Numpy**

#### Pandas can work with numbers and strings together

<!-- - If our **data has only numbers**, we are better off using **Numpy**
  - It's **lighter** and **easier**

- But if our data has both **number and non-number vals**, it makes sense to use **Pandas** -->

So lets see how we can use pandas



## Imagine that you are a Data Scientist with McKinsey



  - McKinsey wants to understand the relation between **GDP per capita** and **life expectancy** and various trends for their clients.
  - The company has acquired **data from multiple surveys** in different countries in the past
  - This contains info of several years about:
   - country
   - population size
   - life expectancy
   - GDP per Capita
  - We have to analyse the data and draw **inferences** meaningful to the company

## Reading dataset in Pandas

Link:https://drive.google.com/file/d/1E3bwvYGf1ig32RmcYiWc0IXPN-mD_bI_/view?usp=sharing

In [2]:
#!wget "https://drive.google.com/uc?export=download&id=1E3bwvYGf1ig32RmcYiWc0IXPN-mD_bI_" -O mckinsey.csv

#### Now how should we read this dataset?

Pandas makes it very easy to work with these kinds of files



In [3]:
df = pd.read_csv('../mckinsey.csv') # We are storing the data in df
df

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

## Dataframe and Series


#### What can we observe from the above dataset ?

We can see that it has:
  - 6 columns
  - 1704 rows

#### What do you think is the datatype of `df` ?

In [4]:
type(df)

pandas.core.frame.DataFrame

Its a **pandas DataFrame**

### What is a pandas DataFrame ?

  - It is a table-like representation of data in Pandas => Structured Data
  - **Structured Data** here can be thought of as **tabular data in a proper order**
  - Considered as **counterpart of 2D-Matrix** in Numpy

#### Now how can we access a column, say `country` of the dataframe?

In [5]:
df["country"]

0       Afghanistan
1       Afghanistan
2       Afghanistan
3       Afghanistan
4       Afghanistan
           ...     
1699       Zimbabwe
1700       Zimbabwe
1701       Zimbabwe
1702       Zimbabwe
1703       Zimbabwe
Name: country, Length: 1704, dtype: object

As you can see we get all the values in the column **country**

#### Now what is the data-type of a column?

In [6]:
type(df["country"])

pandas.core.series.Series

Its a **pandas Series**

### What is a pandas Series ?
  - **Series** in Pandas is what a **Vector** is in Numpy

#### What exactly does that mean?

  - It means a Series is a **single column** of **data**

  - **Multiple Series stack together to form a DataFrame**
  

Now we have understood what Series and DataFrames are

#### What if a dataset has 100 rows ... Or 100 columns ?

#### How can we find the datatype, name, total entries in each column ?


In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1704 entries, 0 to 1703
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   country     1704 non-null   object 
 1   year        1704 non-null   int64  
 2   population  1704 non-null   int64  
 3   continent   1704 non-null   object 
 4   life_exp    1704 non-null   float64
 5   gdp_cap     1704 non-null   float64
dtypes: float64(2), int64(2), object(2)
memory usage: 80.0+ KB


df.info() gives a **list of columns** with:

- **Name/Title** of Columns
- **How many non-null values (blank cells)** each column has
- **Type of values** in each column - int, float, etc.

**By default**, it shows **data-type as `object` for anything other than int or float** - Will come back later

#### Now what if we want to see the first few rows in the dataset ?



In [8]:
df.head()

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

It **Prints top 5 rows by default**

We can also **pass in number of rows we want to see** in `head()`



In [10]:
df.head()

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

#### Similarly what if we want to see the last 20 rows ?

In [12]:
df.tail(20) #Similar to head

country  year  population continent  life_exp      gdp_cap
1684    Zambia  1972     4506497    Africa    50.107  1773.498265
1685    Zambia  1977     5216550    Africa    51.386  1588.688299
1686    Zambia  1982     6100407    Africa    51.821  1408.678565
1687    Zambia  1987     7272406    Africa    50.821  1213.315116
1688    Zambia  1992     8381163    Africa    46.100  1210.884633
1689    Zambia  1997     9417789    Africa    40.238  1071.353818
1690    Zambia  2002    10595811    Africa    39.193  1071.613938
1691    Zambia  2007    11746035    Africa    42.384  1271.211593
1692  Zimbabwe  1952     3080907    Africa    48.451   406.884115
1693  Zimbabwe  1957     3646340    Africa    50.469   518.764268
1694  Zimbabwe  1962     4277736    Africa    52.358   527.272182
1695  Zimbabwe  1967     4995432    Africa    53.995   569.795071
1696  Zimbabwe  1972     5861135    Africa    55.635   799.362176
1697  Zimbabwe  1977     6642107    Africa    57.674   685.587682
1698  Zimbabwe  1982     7636524    Africa    60.363   788.855041
1699  Zimbabwe  1987     9216418    Africa    62.351   706.157306
1700  Zimbabwe  1992    10704340    Africa    60.377   693.420786
1701  Zimbabwe  1997    11404948    Africa    46.809   792.449960
1702  Zimbabwe  2002    11926563    Africa    39.989   672.038623
1703  Zimbabwe  2007    12311143    Africa    43.487   469.709298

#### How can we find the shape of the dataframe?

In [13]:
df.shape

(1704, 6)

Similar to Numpy, it gives **No. of Rows and Columns** -- **Dimensions**

Now we know how to do some basic operations on dataframes



But what if we aren't loading a dataset, but want to create our own.

Let's take a subset of the original dataset

In [14]:
df.head(3) # We take the first 3 rows to create our dataframe

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710

## How can we create a DataFrame from scratch?


### Approach 1: Row-oriented


- It takes **2 arguments** - Because DataFrame is **2-dimensional**
    - A **list of rows**
      - Each **row** is packed in a **list `[]`**
      - All rows are packed in an **outside list `[[]]`** - To **pass a list of rows**
    - A **list of column names/labels**

In [15]:
pd.DataFrame([['Afghanistan',1952, 8425333, 'Asia', 28.801, 779.445314 ],
              ['Afghanistan',1957, 9240934, 'Asia', 30.332, 820.853030 ],
              ['Afghanistan',1962, 102267083, 'Asia', 31.997, 853.100710 ]],
             columns=['country','year','population','continent','life_exp','gdp_cap'])

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962   102267083      Asia    31.997  853.100710

#### Can you create a single row dataframe?

In [16]:
pd.DataFrame(['Afghanistan',1952, 8425333, 'Asia', 28.801, 779.445314 ],
             columns=['country','year','population','continent','life_exp','gdp_cap'])

ValueError: Shape of passed values is (6, 1), indices imply (6, 6)

#### Why did this give an error?


- Because we passed in a **list of values**


- `DataFrame()` expects a **list of rows**

In [17]:
pd.DataFrame([['Afghanistan',1952, 8425333, 'Asia', 28.801, 779.445314 ]],
             columns=['country','year','population','continent','life_exp','gdp_cap'])

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314

### Approach 2: Column-oriented


In [18]:
pd.DataFrame({'country':['Afghanistan', 'Afghanistan'], 'year':[1952,1957],
              'population':[842533, 9240934], 'continent':['Asia', 'Asia'],
              'life_exp':[28.801, 30.332], 'gdp_cap':[779.445314, 820.853030]})

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952      842533      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030

We **pass the data** as a **dictionary**

- **Key** is the **Column Name/Label**


- **Value** is the **list of values column-wise**


We now have a basic idea about the dataset and creating rows and columns


What kind of **other operations** can we perform on the dataframe?

Thinking from database perspective:
- Adding data
- Removing data
- Updating/Modifying data

and so on

## Basic operations on columns


Now what operations can we do using columns?
- Maybe add a column
- or delete a column
- or we can rename the column too

and so on.


We can see that our dataset has 6 cols

#### But what if our dataset has 20 cols ? ... or 100 cols ? We can't see ther names in **one go**.

#### How can we get the names of all these cols ?

We can do it in two ways:
  1. df.columns
  2. df.keys



In [19]:
df.columns  # using attribute `columns` of dataframe

Index(['country', 'year', 'population', 'continent', 'life_exp', 'gdp_cap'], dtype='object')

In [20]:
df.keys()  # using method keys() of dataframe

Index(['country', 'year', 'population', 'continent', 'life_exp', 'gdp_cap'], dtype='object')

Note:

- Here, `Index` is a type of pandas class used to store the `address` of the series/dataframe

- It is an Immutable sequence used for indexing and alignment.





In [ ]:
# df['country'].head()  # Gives values in Top 5 rows pertaining to the key

Pandas DataFrame and Series are specialised dictionary

#### But what is so "special" about this dictionary?

It can take multiple keys


In [21]:
df[['country', 'life_exp']].head()

country  life_exp
0  Afghanistan    28.801
1  Afghanistan    30.332
2  Afghanistan    31.997
3  Afghanistan    34.020
4  Afghanistan    36.088

And what if we pass a single column name?

In [22]:
df[['country']].head()

country
0  Afghanistan
1  Afghanistan
2  Afghanistan
3  Afghanistan
4  Afghanistan

In [23]:
df['country'].head()

0    Afghanistan
1    Afghanistan
2    Afghanistan
3    Afghanistan
4    Afghanistan
Name: country, dtype: object

Note:

Notice how this output type is different from our earlier output using `df['country']`

==> `['country']` gives series while `[['country']]` gives dataframe

Now that we know how to access columns, lets answer some questions

### How can we find the countries that have been surveyed ?

We can find the unique vals in the `country` col

#### How can we find unique values in a column?

In [24]:
df['country'].unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Angola', 'Argentina',
       'Australia', 'Austria', 'Bahrain', 'Bangladesh', 'Belgium',
       'Benin', 'Bolivia', 'Bosnia and Herzegovina', 'Botswana', 'Brazil',
       'Bulgaria', 'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon',
       'Canada', 'Central African Republic', 'Chad', 'Chile', 'China',
       'Colombia', 'Comoros', 'Congo, Dem. Rep.', 'Congo, Rep.',
       'Costa Rica', "Cote d'Ivoire", 'Croatia', 'Cuba', 'Czech Republic',
       'Denmark', 'Djibouti', 'Dominican Republic', 'Ecuador', 'Egypt',
       'El Salvador', 'Equatorial Guinea', 'Eritrea', 'Ethiopia',
       'Finland', 'France', 'Gabon', 'Gambia', 'Germany', 'Ghana',
       'Greece', 'Guatemala', 'Guinea', 'Guinea-Bissau', 'Haiti',
       'Honduras', 'Hong Kong, China', 'Hungary', 'Iceland', 'India',
       'Indonesia', 'Iran', 'Iraq', 'Ireland', 'Israel', 'Italy',
       'Jamaica', 'Japan', 'Jordan', 'Kenya', 'Korea, Dem. Rep.',
       'Korea, Rep.', 'Kuwait', 'Leba

#### Now what if you also want to check the count of each country in the dataframe?


In [25]:
df['country'].value_counts()

Afghanistan          12
Pakistan             12
New Zealand          12
Nicaragua            12
Niger                12
                     ..
Eritrea              12
Equatorial Guinea    12
El Salvador          12
Egypt                12
Zimbabwe             12
Name: country, Length: 142, dtype: int64

Note:

`value_counts()` shows the output in **decreasing order of frequency**

### What if we want to change the name of a column ?

We can rename the column by:
- passing the dictionary with `old_name:new_name` pair
- specifying `axis=1`

In [32]:
df.rename({"population": "Population", 
           "country":"Country" }, 
          axis = 1)

Country  year  Population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

Alternatively, we can also rename the column without using `axis`
- by using the `column` parameter



In [33]:
df.rename(columns={"country":"Country"})

Country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

We can set it inplace by setting the `inplace` argument = True

In [34]:
df.rename({"country": "Country"}, axis = 1, inplace = True)
df

Country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

**Note**
- .rename has default value of axis=0
- If two columns have the **same name**, then `df['column']` will display both columns

Now lets try another way of accessing column vals

In [ ]:
df.Country

0       Afghanistan
1       Afghanistan
2       Afghanistan
3       Afghanistan
4       Afghanistan
           ...     
1699       Zimbabwe
1700       Zimbabwe
1701       Zimbabwe
1702       Zimbabwe
1703       Zimbabwe
Name: Country, Length: 1704, dtype: object

This however doesn't work everytime

#### What do you think could be the problems with using attribute style for accessing the columns?

**Problems** such as
- if the column names are **not strings**
  - Starting with **number**: E.g., ```2nd```
  - Contains a **space**: E.g., ```Roll Number```
- or if the column names conflict with **methods of the DataFrame**
  - E.g. ```shape```

It is generally better to avoid this type of accessing columns

#### Are all the columns in our data necessary?
  
  - We already know the continents in which each country lies
  - So we don't need this column


### How can we delete cols in pandas dataframe ?


In [35]:
df.drop('continent', axis=1)

Country  year  population  life_exp     gdp_cap
0     Afghanistan  1952     8425333    28.801  779.445314
1     Afghanistan  1957     9240934    30.332  820.853030
2     Afghanistan  1962    10267083    31.997  853.100710
3     Afghanistan  1967    11537966    34.020  836.197138
4     Afghanistan  1972    13079460    36.088  739.981106
...           ...   ...         ...       ...         ...
1699     Zimbabwe  1987     9216418    62.351  706.157306
1700     Zimbabwe  1992    10704340    60.377  693.420786
1701     Zimbabwe  1997    11404948    46.809  792.449960
1702     Zimbabwe  2002    11926563    39.989  672.038623
1703     Zimbabwe  2007    12311143    43.487  469.709298

[1704 rows x 5 columns]

The `drop` function takes two parameters:
  - The column name
  - The axis
  
By default the value of `axis` is 0



An alternative to the above approach is using the "columns" parameter as we did in rename

In [36]:
df.drop(columns=['continent'])

Country  year  population  life_exp     gdp_cap
0     Afghanistan  1952     8425333    28.801  779.445314
1     Afghanistan  1957     9240934    30.332  820.853030
2     Afghanistan  1962    10267083    31.997  853.100710
3     Afghanistan  1967    11537966    34.020  836.197138
4     Afghanistan  1972    13079460    36.088  739.981106
...           ...   ...         ...       ...         ...
1699     Zimbabwe  1987     9216418    62.351  706.157306
1700     Zimbabwe  1992    10704340    60.377  693.420786
1701     Zimbabwe  1997    11404948    46.809  792.449960
1702     Zimbabwe  2002    11926563    39.989  672.038623
1703     Zimbabwe  2007    12311143    43.487  469.709298

[1704 rows x 5 columns]

As you can see, **column `contintent` is dropped**

#### Has the column permanently been deleted?

In [37]:
df.head()

Country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

NO, the **column `continent` is still there**

**Do you see what's happening here?**

We only got a **view of dataframe with column `continent` dropped**

#### How can we permanently drop the column?

We can either **re-assign** it
- `df = df.drop('continent', axis=1)`

  OR
    
- We can **set parameter `inplace=True`**
    
By **default, `inplace=False`**

In [38]:
df.drop('continent', axis=1, inplace=True)

In [39]:
df.head() #we print the head to check

Country  year  population  life_exp     gdp_cap
0  Afghanistan  1952     8425333    28.801  779.445314
1  Afghanistan  1957     9240934    30.332  820.853030
2  Afghanistan  1962    10267083    31.997  853.100710
3  Afghanistan  1967    11537966    34.020  836.197138
4  Afghanistan  1972    13079460    36.088  739.981106

Now we can see the column `continent` is permanently dropped

### Now similarly, what if we want to create a new column?

We can either
- use values from **existing columns**

OR
- create our **own values**

#### How to create a column using values from an existing column?

In [40]:
df["year+7"] = df["year"] + 7
df.head()

Country  year  population  life_exp     gdp_cap  year+7
0  Afghanistan  1952     8425333    28.801  779.445314    1959
1  Afghanistan  1957     9240934    30.332  820.853030    1964
2  Afghanistan  1962    10267083    31.997  853.100710    1969
3  Afghanistan  1967    11537966    34.020  836.197138    1974
4  Afghanistan  1972    13079460    36.088  739.981106    1979

As we see, a new column `year+7` is created from the column `year`

We can also use values from two columns to form a new column

#### Which two columns can we use to create a new column `gdp`?

In [41]:
df['gdp']=df['gdp_cap'] * df['population']
df.head()

Country  year  population  life_exp     gdp_cap  year+7           gdp
0  Afghanistan  1952     8425333    28.801  779.445314    1959  6.567086e+09
1  Afghanistan  1957     9240934    30.332  820.853030    1964  7.585449e+09
2  Afghanistan  1962    10267083    31.997  853.100710    1969  8.758856e+09
3  Afghanistan  1967    11537966    34.020  836.197138    1974  9.648014e+09
4  Afghanistan  1972    13079460    36.088  739.981106    1979  9.678553e+09

#### As you can see

- An **additional column** has been **created**

- **Values** in this column are **product of respective values in `gdp_cap` and `population`**


#### What other operations we can use?

Subtraction, Addition, etc.

### How can we create a new column from our own values?

- We can **create a list**

OR

- We can **create a Pandas Series** from a list/numpy array for our new column

In [43]:
df["Own"] = [i for i in range(df.shape[0])]  # count of these values should be correct
df

Country  year  population  life_exp     gdp_cap  year+7  \
0     Afghanistan  1952     8425333    28.801  779.445314    1959   
1     Afghanistan  1957     9240934    30.332  820.853030    1964   
2     Afghanistan  1962    10267083    31.997  853.100710    1969   
3     Afghanistan  1967    11537966    34.020  836.197138    1974   
4     Afghanistan  1972    13079460    36.088  739.981106    1979   
...           ...   ...         ...       ...         ...     ...   
1699     Zimbabwe  1987     9216418    62.351  706.157306    1994   
1700     Zimbabwe  1992    10704340    60.377  693.420786    1999   
1701     Zimbabwe  1997    11404948    46.809  792.449960    2004   
1702     Zimbabwe  2002    11926563    39.989  672.038623    2009   
1703     Zimbabwe  2007    12311143    43.487  469.709298    2014   

               gdp   Own  
0     6.567086e+09     0  
1     7.585449e+09     1  
2     8.758856e+09     2  
3     9.648014e+09     3  
4     9.678553e+09     4  
...            ...   ...  
1699  6.508241e+09  1699  
1700  7.422612e+09  1700  
1701  9.037851e+09  1701  
1702  8.015111e+09  1702  
1703  5.782658e+09  1703  

[1704 rows x 8 columns]

Now that we know how to create new cols lets see some basic ops on rows

Before that lets drop the newly created cols


In [44]:
df.drop(columns=["Own",'gdp', 'year+7'], axis = 1, inplace = True)
df

Country  year  population  life_exp     gdp_cap
0     Afghanistan  1952     8425333    28.801  779.445314
1     Afghanistan  1957     9240934    30.332  820.853030
2     Afghanistan  1962    10267083    31.997  853.100710
3     Afghanistan  1967    11537966    34.020  836.197138
4     Afghanistan  1972    13079460    36.088  739.981106
...           ...   ...         ...       ...         ...
1699     Zimbabwe  1987     9216418    62.351  706.157306
1700     Zimbabwe  1992    10704340    60.377  693.420786
1701     Zimbabwe  1997    11404948    46.809  792.449960
1702     Zimbabwe  2002    11926563    39.989  672.038623
1703     Zimbabwe  2007    12311143    43.487  469.709298

[1704 rows x 5 columns]